Adding documentation for matProps (#2519)
terrapower/armi: doc/user/inputs.rst

  different pressure loss coefficients and/or flow rates to different types of assemblies.

material modifications
  There are a variety of material modifications available for each material. The most common material modifications are
  usually to do with enrichment of a particular nuclide, modifying the theoretical density of a material, or changing

>   some list of custom isotpics. You can set values to these material modifications in your blueprints. And if you want

Reviewer: ```suggestion
  some list of custom isotopics. You can set values to these material modifications in your blueprints. And if you want
```

  to add new modifications to your own custom material, you would typically implement that change in your material's
  ``applyInputParams()`` method. Here are some example material modifications from ARMI's history:

  * Lithium: LI6_wt_frac
  * B4C: B10_wt_frac, TD_frac

>   * Uranium: U235_wt_frac, TD_frac, class1_custom_isotopics, class2_custom_isotopics, class1_wt_frac, customIsotopics

Reviewer: What would be the use case for `class1_wt_frac` if `U235_wt_frac` is an example already?

And is `class1_custom_isotopics` used as such or is there a descriptor someone would use for `class1`, etc? 

Is `customIsotopics` supported still?

Author: I don't know what `class1_wt_frac` was ever used for.  It appears to be unused everywhere I can find, except one  equilibrium unit test. Should we make a ticket to remove it?

Though `customIsotopics` is still used, yeah.

Reviewer: well I think `class1` is always replaced with a nuclide in practice, And I assumed `class1` would be replaced with a descriptor when used with `_custom_isotopics` as well, but I haven't seen anyone use it

Author: Honestly, it's a feature I'd have to test and see if anyone uses the hard way: make a branch where we remove the logic, run all the downstream tests, and mini runtime cases, we can and see if anything breaks.

Anything that is string-replacement like this gets hard to just `grep` for to see if it's used.

Reviewer: Yea I agree. It is from what I can tell an unused feature (doesn't mean it should be removed though, but maybe the `classN_custom_isotopics` could be removed from this particular list).

The `_wt_frac` example here could be cleared up. I think including `class1_wt_frac` confuses things. The code uses that, but the user uses `<NUC>_wt_frac`.

Author: Removed

  In the Lithium example above, both material modifications modify the weight fraction, or mass fraction, of an element
  or nuclide. That is useful if you need to specify the fraction of a particular nuclide in your material. Similarly,
  the B4C, and Uranium examples above have things like "B10_wt_frac" and "U235_wt_frac" where the goal is
  clearly to set the mass enrichment of some important nuclide in your material. This is particularly common in
  depletable fuel-type fuels.

  A popular class of material modifications is adjusting the theoretical density of a material. The modification name

>   for this is usually "TD_frac" as a soft convention. This is popular for solids when the actual density of the

Reviewer: Is it a soft convention? I thought it was ARMI's regular firmness convention. Or you're saying it's the convention we chose but one can implement an `applyInputParams` method to take in anything?

Author: Well, ARMI would like to pretend it's a hard convention. 

But if you create a new Python material, you can write a new `applyInputParams` method that takes in parameters named anything you like. No one can stop you. 

That's all I meant.

Reviewer: ah ok. We can leave it as is I think

  material is slightly different than the theoretical density due to the manufacturing process.

  The class 1/class 2 modifications in fuel materials are used to identify mixtures of custom isotopics labels for

>   input scenarios where a varying blend of a high-reactivity feed with a low-reactivity feed. This is often useful for

Reviewer: ```suggestion
  input scenarios where there is a blend of a high-reactivity feed with a low-reactivity feed. This is often useful for
```

  closed fuel cycles. For example, you can define any fuel material as being made of LWR-derived TRU plus depleted
  uranium at various weight fractions. Note that this input style only adjusts the heavy metal.

>   User can specify material modications on a by-component basis. This is useful, for instance, when two pins within an

Reviewer: ```suggestion
  Users can specify material modications on a by-component basis. This is useful, for instance, when two pins within an
```

  different pressure loss coefficients and/or flow rates to different types of assemblies.

material modifications
  There are a variety of material modifications available for each material. The most common material modifications are
  usually to do with enrichment of a particular nuclide, modifying the theoretical density of a material, or changing
  some list of custom isotpics. You can set values to these material modifications in your blueprints. And if you want

>   to add new modifications to your own custom material, you would typically implement that change in your materials's

Reviewer: ```suggestion
  to add new modifications to your own custom material, you would typically implement that change in your material's
```

Author: done

>   `applyInputParams()` method. Here are some example materical modifications from ARMI's history:

Reviewer: ```suggestion
  ``applyInputParams()`` method. Here are some example material modifications from ARMI's history:
```

Author: done

  * Lithium: LI6_wt_frac

>   * Sulfur: sulfur_density_frac, TD_frac

Reviewer: `sulfur_density_frac`: Is this used at all, or is it just `TD_frac` now?

Author: Sure, I just removed the sulfur example entirely.

>   * B4C: B10_wt_frac, theoretical_density, TD_frac

Reviewer: ```suggestion
  * B4C: B10_wt_frac, TD_frac
```

Author: done

  * Uranium: U235_wt_frac, TD_frac, class1_custom_isotopics, class2_custom_isotopics, class1_wt_frac, customIsotopics

  In the Lithium example above, both material modifications modify the weight fraction, or mass fraction, of an element
  or nuclide. That is useful if you need to specify the fraction of a particular nuclide in your material. Similarly,
  the Sulfur, B4C, and Uranium examples above have things like "B10_wt_frac" and "U235_wt_frac" where the goal is
  clearly to set the mass enrichment of some important nuclide in your material. This is particularly common in
  depletable fuel-type fuels.

  A popular class of material modifications is adjusting the theoretical density of a material. The modification name
  for this is usually "TD_frac" as a soft convention. But you will also see "theoretical_density" above, and can pick

>   whatever naming convention you like. This is popular for solids when the actual density of the material is slightly

Reviewer: ```suggestion
  for this is "TD_frac". This is popular for solids when the actual density of the material is slightly
```
`theoretical_density` is deprecated. We actually can remove it now/whenever. Nowhere downstream uses it.

Author: done

    defProp("goo", "gooiness", "m^2/s")
    defProp("squish", "squishiness", "1/Pa")

    armi.matProps.loadSafe("path/to/hilarious/materials/")

Tabulated Data
""""""""""""""
A material property can be defined not just by math, but by tabulated data. This is flexible, easy to use, and can
better match laboratory measurements. Currently, matProps supports one dimensional and two dimensional tables of data,
as shown in the example at the beginning of this section. Notice that the "type" field is used to make this distinction.

These tables of data can be large and inconvienent to duplicate. So you can tag and name the tables of data for use
elsewhere in the YAML file. (Look for ``"tabulated data: &tagged_data"`` in the example file above.)

Functions
"""""""""
The ``function`` field allows the developer to define a mathematical curve for a material property. For instance, in the
example file above, a function is used to represent the "Young's modulus" of the material. The independent variable "T"
is defined to represent Temperature in degrees C, and the min and max values of 0 and 700 are given. This defines the
valid bounds upon which the polynomial "equation" can be evaluated.

> The ``function`` system is very flexible though. Notice that "Poisson's ratio" above has "equation: 0.321", so constant

Reviewer: ```suggestion
The ``function`` system is very flexible though. Notice that "Poisson's ratio" above has ``"equation: 0.321"``, so constant
```

    | cosh        | Hyperbolic Cosine                |
    +-------------+----------------------------------+
    | tanh        | Hyperbolic Tangent               |
    +-------------+----------------------------------+
    | ln          | Natural Logarithm                |
    +-------------+                                  +
    | log         |                                  |
    +-------------+----------------------------------+
    | log10       | Base 10 Logarithm                |
    +-------------+----------------------------------+
    | exp         | Exponential (base e)             |
    +-------------+----------------------------------+
    | e or E      | Scientific Notation              |
    +-------------+----------------------------------+

Table
=====
Providing ``table`` for the ``type`` key indicates a one dimensional tables that uses interpolation. This function type
requires a set of tabulated data to be defined in the collection. The set of tabulated data is a list of lists of length
2 with the first element being the independent variable, and the second element being the property value at that
independent variable value. 

Piecewise
=========

> Using the "piecewise" keyword allows multiple functions to be defined for different ranges of independent variable

Reviewer: In this section, you switch from using double backtick formatting to quote formatting when describing the field. Should have it be consistent (I don't care which way you go, just that it's consistent)

Author: fixed

values. As in the "density" example above, the field "functions" is used to define a list of functions, one for each
range defined.

.. note:: While piecewise-defined functions are allowed to have gaps in the valid range between child functions, the
    child function valid regions may not overlap. An error will be raised during parsing the set of child functions have
    overlapping valid ranges or utilize different independent variables.

.. note:: If the piecewise-defined function is discontinuous at a point (two child functions overlap at a point), the
    function that is defined first in the input file will be used. 

Two Dimensional Table
=====================
The "two dimensional table" field defines a two-dimensional table, that needs to be supplied a specially formatted data
set. See the "stress to rupture" property in the above example file. This can again use the tagged data system to share

> data with other fields, see ``&tagged_data`` in the above example file.

Reviewer: ```suggestion
data with other fields, see ``"&tagged_data"`` in the above example file.
```
Just to align with the prior formatting

.. |Tinput| replace:: T\ :sub:`input`
.. |Thot| replace:: T\ :sub:`hot`

The Materials Input File
========================
The **materials** input files define materials to be used during an ARMI-based simulation. The YAML data format below is
the preferred way to define materials in ARMI, and they are loaded by the ``armi.matProps`` package. However, for
increased flexiblity, the ``armi.materials`` package allows users to define materials entirely in Python, without
dealing with the below YAML data format. Please see the ``armi.materials.material.Material`` class for more information.

You can find example YAML material files under ``armi/resources/materials/`` and example Python materials under:
``armi/materials/``.

This YAML material file format is human readable, hierarchical, and compact. The base name of the YAML filename (the

> file name without the file path or extension suffix) is interpreted by matProps as the case-sensitive material name. The

Reviewer: ```suggestion
file name without the file path or extension suffix) is interpreted by ``matProps`` as the case-sensitive material name. The
```

extension for the material data file must be one of these: ``.yaml``, ``.yml``, ``.YAML``, or ``.YML``.

> Below is an example material data file containing all the basic matProps concepts, discussed in the following sections.

Reviewer: ```suggestion
Below is an example material data file containing all the basic ``matProps`` concepts, discussed in the following sections.
```

verified to see if it is supported for use by ``matProps``.

.. _ref_matType:

Material Type 
"""""""""""""
The ``material type`` field defines the type of material. Valid values for this key include Metal, Fuel, Fluid, Ceramic,
ASME2015, ASME2017, SimpleSolid, and Composite. This is a required field. This field is meant to provide information to
the user and is not used by ``matProps``, though it may be used by downstream codes.

.. _ref_comp:

Composition
"""""""""""
The ``composition`` field defines the chemical composition of the material. This is a required field. The value for this
field is a collection of key-value pairs, as denoted by indentation. The key is an element name,
and the value is a list of length 2 that defines the minimum percent composition then the maximum percent composition
for the element. There must be one and only one element listed as "balance" instead of a pair of minimum and maximum values. The "balance"  element will fill any remaining fraction of the elements composition, to get the total to
100 percent.

Optionally, this data entry can be associated with a references keyword.

Composition Requirements:

> - The value ``balance`` shall be assigned to one and only one element.

Reviewer: ```suggestion
- The value "balance" shall be assigned to one and only one element.
```
Just for uniformity with above formatting.

material, like density. The value for this field is a collection consisting of several required and optional keywords. A
summary of the keywords, their relevant status and appropriate section can be found below.

.. table:: Summary of Property Components

    +----------------------+-------------------------------------+--------------------------+
    | Property Components  | Required Status                     | Section                  |
    +======================+=====================================+==========================+
    | function             | Required                            | :numref:`ref_functions`  |
    +----------------------+-------------------------------------+--------------------------+
    | references           | Optional                            | :numref:`ref_references` |
    +----------------------+-------------------------------------+--------------------------+
    | tabulated data       | Required for table data functions.  | :numref:`ref_tabdata`    |
    |                      | Otherwise, it is optional.          |                          |
    +----------------------+-------------------------------------+--------------------------+

> The table below lists the default collection of properties that are defined for a material by matProps. 

Reviewer: ```suggestion
The table below lists the default collection of properties that are defined for a material by ``matProps``. 
```

    +------------------------------------------------+-------------------+
    | fracture toughness                             | ``K_IC``          |
    +------------------------------------------------+-------------------+
    | Brinell Hardness                               | ``HBW``           |
    +------------------------------------------------+-------------------+
    | factor f from ASME.III.5 Fig. HBB-T-1432-2     | ``f``             |
    +------------------------------------------------+-------------------+
    | factor Kv' from ASME.III.5 Fig. HBB-T-1432-3   | ``Kv_prime``      |
    +------------------------------------------------+-------------------+

It is quite easy to add another material property to your simulation. Use the ``defProp`` function, and describe the new
property by: short name, long name, and units. It will then be available to use in your material YAML files like any
other property.

.. code-block:: python

    from armi.matProps.prop import defProp

    defProp("fuzz", "fuzziness", "1/m^2")
    defProp("goo", "gooiness", "m^2/s")
    defProp("squish", "squishiness", "1/Pa")

>     armi.matProps.loadSafe("path/to/hilarious/materials/")

Author: ```suggestion
    armi.matProps.loadSafe("/pop/six/squish/uhuh/cicero/lipschitz")
```

>     armi.matProps.loadSafe("path/to/hilarious/materials/")

Reviewer: would it be copyright infringement to have the filepath be `/pop/six/squish/uhuh/cicero/lipschitz` bc that's all I can think about when I read the material properties above.....

Reviewer: 😁

Tabulated Data
""""""""""""""
A material property can be defined not just by math, but by tabulated data. This is flexible, easy to use, and can
better match laboratory measurements. Currently, ``matProps`` supports one dimensional and two dimensional tables of

> data, as shown in the example at the beginning of this section. Notice that the "type" field is used to make this

Reviewer: ```suggestion
data, as shown in the example at the beginning of this section. Notice that the ``type`` field is used to make this
```

distinction.

These tables of data can be large and inconvienent to duplicate. So you can tag and name the tables of data for use
elsewhere in the YAML file. (Look for ``"tabulated data: &tagged_data"`` in the example file above.)

Functions
"""""""""
The ``function`` field allows the developer to define a mathematical curve for a material property. For instance, in the
example file above, a function is used to represent the "Young's modulus" of the material. The independent variable "T"
is defined to represent Temperature in degrees C, and the min and max values of 0 and 700 are given. This defines the
valid bounds upon which the polynomial "equation" can be evaluated.

The ``function`` system is very flexible though. Notice that "Poisson's ratio" above has ``"equation: 0.321"``, so constant
functions are easy to define. And "yield strength" is actually a function of two independent variables, which
``matProps`` supports. Functions can also include tabular data. For tabular functions, instead of child keys, the value
``0`` or ``1`` is provided to indicate which dimension of the table corresponds to which variable. Values returned from

> tabular functions will be interpretted from the provided values. Several other types of functions are defined in detail

Reviewer: ```suggestion
tabular functions will be interpreted from the provided values. Several other types of functions are defined in detail
```

    | tanh        | Hyperbolic Tangent               |
    +-------------+----------------------------------+
    | ln          | Natural Logarithm                |
    +-------------+                                  +
    | log         |                                  |
    +-------------+----------------------------------+
    | log10       | Base 10 Logarithm                |
    +-------------+----------------------------------+
    | exp         | Exponential (base e)             |
    +-------------+----------------------------------+
    | e or E      | Scientific Notation              |
    +-------------+----------------------------------+

Table
=====
Providing ``table`` for the ``type`` key indicates a one dimensional tables that uses interpolation. This function type
requires a set of tabulated data to be defined in the collection. The set of tabulated data is a list of lists of length
2 with the first element being the independent variable, and the second element being the property value at that
independent variable value. 

Piecewise
=========
Using the ``piecewise`` keyword allows multiple functions to be defined for different ranges of independent variable

> values. As in the ``density`` example above, the field "functions" is used to define a list of functions, one for each

Reviewer: ```suggestion
values. As in the density example above, the field ``functions`` is used to define a list of functions, one for each
```

Reviewer: I'm trying to make fields monospace and properties not. We will see how successful I am

share data with other fields, see ``"&tagged_data"`` in the above example file.

The form of the inner lists varies depending on whether the list corresponds to the first row or one of the subsequent
rows. The first row will always have its first entry be a ``None`` value. The second value will be a list containing all
of the values for the first independent variable (indicated by ``0`` in the independent variable node), in ascending
order, that are utilized in the two-dimensional table. Each subsequent row will have its first entry be the second
independent variable value corresponding to that row. The second entry will be the property values that map to the
independent variable combination for that row and column. The size of the lists in the second entry must be consistent
for each row. If there is a property value that does not exist for an independent variable combination, then a value of
``None`` must be specified in its place. If the point including ``None`` is used in evaluation, an error will be raised.

Troubleshooting
---------------
The ``armi.matProps`` package includes error checking to avoid invalid states or returning inaccurate data. Any error

> found raises an ``Exception`` that includes a helpful message. The four broad categories of matProps errors are:

Reviewer: ```suggestion
found raises an ``Exception`` that includes a helpful message. The four broad categories of ``matProps`` errors are:
```

.. |Tinput| replace:: T\ :sub:`input`
.. |Thot| replace:: T\ :sub:`hot`

The Materials Input File
========================
The **materials** input files define materials to be used during an ARMI-based simulation. The YAML data format below is

> the preferred way to define materials in ARMI, and they are loaded by the `armi.matProps` package. However, for

Reviewer: I think all the single backticks here need to be double backticks in here

Reviewer: Yep, confirmed. All single backtick entries render as italic and not monospace.

Author: Damn, sorry. Done.

                max: 5000
            type: symbolic
            equation: 0.321

    yield strength:
        function:
            T: 
                min: 350
                max: 500
            D: 
                min: 0
                max: 100
            type: symbolic
            equation: 400.0 * T**2.0 + D

    stress to rupture:
        function:
            T: 0
            t: 1
            type: two dimensional table
        tabulated data:
          - [null,    [   100.,    200.,    300.]]
          - [1000.,   [123.456, 23456.7, 456.789]]
          - [10000.,  [78910.1, 89.0123, 10111.2]]

    design fatigue strain range:
        function:
            T: 0
            n: 1
            type: two dimensional table
        tabulated data:
            - [null,     [  427.,   600.,   775.,   800.]]
            - [10.,      [  0.012, 0.0059,0.0036, 0.0029]]
            - [20.,      [  0.01,  0.0051,0.0033, 0.0020]]

The only required entries in a material file are "file format", "composition", and "material type".

File Format 
"""""""""""
The `file format: version` field defines the file format version. This is a required field. This version string is

> verified to see if it is supported for use by `matProps`.

Reviewer: we have some standalone matProps and some code-formatted `matProps` throughout this text. I say pick one and make it consistent, and I don't have a strong opinion on which way you go

Author: I moved them all from matProps to `matProps`.

.. _ref_matType:

Material Type 
"""""""""""""
The `material type` field defines the type of material. Valid values for this key include Metal, Fuel, Fluid, Ceramic,
ASME2015, ASME2017, SimpleSolid, and Composite. This is a required field. This field is meant to provide information to

> the user, and is not used by matProps. Though it may be used by downstream codes.

Reviewer: ```suggestion
the user and is not used by matProps, though it may be used by downstream codes.
```

.. _ref_comp:

Composition
"""""""""""
The `composition` field defines the chemical composition of the material. This is a required field. The value for this
field is a collection of key-value pairs, as denoted by indentation. In this second layer, the key is an element name,
and the value is a list of length 2 that defines the minimum percent composition then the maximum percent composition

> for the element. There must be one and only one element’s value listed as "balance" instead of a pair of minimum and maximum values. The "balance"  element will fill any remaining fraction of the elements composition, to get the total to

Reviewer: ```suggestion
field is a collection of key-value pairs, as denoted by indentation. The key is an element name,
and the value is a list of length 2 that defines the minimum percent composition then the maximum percent composition
for the element. There must be one and only one element listed as "balance" instead of a pair of minimum and maximum values. The "balance"  element will fill any remaining fraction of the elements composition, to get the total to
```

summary of the keywords, their relevant status and appropriate section can be found below.

.. table:: Summary of Property Components

    +----------------------+-------------------------------------+--------------------------+
    | Property Components  | Required Status                     | Section                  |
    +======================+=====================================+==========================+
    | function             | Required                            | :numref:`ref_functions`  |
    +----------------------+-------------------------------------+--------------------------+
    | references           | Optional                            | :numref:`ref_references` |
    +----------------------+-------------------------------------+--------------------------+
    | tabulated data       | Required for table data functions.  | :numref:`ref_tabdata`    |
    |                      | Otherwise, it is optional.          |                          |
    +----------------------+-------------------------------------+--------------------------+

The table below lists the default collection of properties that are defined for a material by matProps. Though this list

> is easy to expand. Note that these property names are case sensitive.

Reviewer: ```suggestion
The table below lists the default collection of properties that are defined for a material by matProps. 
Note that these property names are case sensitive.
```

.. table:: Available material properties for YAML definition and corresponding matProps API.

    +------------------------------------------------+---------------+

>     | Material Property YAML Name                    | API Name      |

Reviewer: Should the API Name column contents get double backticks for monospace formatting in this table?

Author: Done

    +------------------------------------------------+---------------+
    | Brinell Hardness                               | HBW           |
    +------------------------------------------------+---------------+
    | factor f from ASME.III.5 Fig. HBB-T-1432-2     | f             |
    +------------------------------------------------+---------------+
    | factor Kv' from ASME.III.5 Fig. HBB-T-1432-3   | Kv_prime      |
    +------------------------------------------------+---------------+

It is quite easy to add another material property to your simulation. Use the `defProp` function, and describe the new
property by: short name, long name, and units. It will then be available to use in your material YAML files like any
other property.

.. code-block:: python

    from armi.matProps.prop import defProp

    defProp("fuzz", "fuzziness", "1/m^2")
    defProp("goo", "gooiness", "m^2/s")
    defProp("squish", "squishiness", "1/Pa")

    armi.matProps.loadSafe("path/to/hilarious/materials/")

Tabulated Data
""""""""""""""
A material property can be defined not just by math, but by tabulated data. This is flexible, easy to use, and can

> better match laboratory measurements. Currently, matProps supports one dimensional and two dimensional tables of data,

Reviewer: Is this limitation still true?

Author: Yup. You will see two files under `armi.matProps`: `1DtableFunction.py` and `2DtableFunction.py`.

 That logic is not generalized. Or not yet. We could make a ticket.

Reviewer: AH ok I just had a lapse in memory while reviewing

as shown in the example at the beginning of this section. Notice that the "type" field is used to make this distinction.

These tables of data can be large and inconvienent to duplicate. So you can tag and name the tables of data for use

> elsewhere in the YAML file. (Look for "tabulated data: &tagged_data" in the example file above.)

Reviewer: ```suggestion
elsewhere in the YAML file. (Look for ``tabulated data: &tagged_data`` in the example file above.)
```

Functions
"""""""""
The `function` field allows the developer to define a mathematical curve for a material property. For instance, in the
example file above, a function is used to represent the "Young's modulus" of the material. The independent variable "T"
is defined to represent Temperature in degrees C, and the min and max values of 0 and 700 are given. This defines the

> valid bounds upon which the "equation" can be evaluated. And the "equation" listed is a polynomial equation.

Reviewer: ```suggestion
valid bounds upon which the polynomial "equation" can be evaluated.
```

Perhaps I'm missing some context with this suggestion

The `function` system is very flexible though. Notice that "Poisson's ratio" above has "equation: 0.321", so constant
functions are easy to define. And "yield strength" is actually a function of two independent variables, which matProps
supports. Functions can also include tabular data. For tabular functions, instead of child keys, the value `0` or `1` is

> provided to indicate which dimension of the table corresponds to which variable. Values returned from tabular functions will be interpretted from the provided values. Several other types of functions are defined in detail below.

Reviewer: ```suggestion
provided to indicate which dimension of the table corresponds to which variable. Values returned from tabular functions will be interpreted from the provided values. Several other types of functions are defined in detail below.
```

Like most everything else in a matProps YAML file, this field can (and probably should) also include a list of
references to properly attribute the source of the data.

Symbolic
========
A symbolic function is defined by supplying value `symbolic` for the key `type`. Symbolic functions may use any number
of independent variables. The `equation` node must be supplied which contains a string with the equation function.
The `symbolicOperators` field defines the set of operators that may be used in the symbolic equation string. These

> operators may be combined in any order or grouping in order to build up a symbolic expression. While other operators

Reviewer: "or grouping in order" is confusing to me.

Author: I just removed that.

    +-------------+                                  +
    | log         |                                  |
    +-------------+----------------------------------+
    | log10       | Base 10 Logarithm                |
    +-------------+----------------------------------+
    | exp         | Exponential (base e)             |
    +-------------+----------------------------------+
    | e or E      | Scientific Notation              |
    +-------------+----------------------------------+

Table
=====
Providing `table` for the `type` key indicates a one dimensional tables that uses interpolation. This function type
requires a set of tabulated data to be defined in the collection. The set of tabulated data is a list of lists of length
2 with the first element being the independent variable, and the second element being the property value at that
independent variable value. 

Piecewise
=========
Using the "piecewise" keyword allows multiple functions to be defined for different ranges of independent variable
values. As in the "density" example above, the field "functions" is used to define a list of functions, one for each
range defined.

> .. note:: While, piecewise-defined functions are allowed to have gaps in the valid range between child functions, the

Reviewer: ```suggestion
.. note:: While piecewise-defined functions are allowed to have gaps in the valid range between child functions, the
```

    child function valid regions may not overlap. An error will be raised during parsing the set of child functions have
    overlapping valid ranges or utilize different independent variables.

.. note:: If the piecewise-defined function is discontinuous at a point (two child functions overlap at a point), the
    function that is defined first in the input file will be used. 

Two Dimensional Table
=====================
The "two dimensional table" field defines a two-dimensional table, that needs to be supplied a specially formatted data
set. See the "stress to rupture" property in the above example file. This can again use the tagged data system to share

> data with other fields, see the "&tagged_data" in the above example file.

Reviewer: ```suggestion
data with other fields, see ``&tagged_data`` in the above example file.
```

The form of the inner lists varies depending on whether the list corresponds to the first row or one of the subsequent
rows. The first row will always have its first entry be a `None` value. The second value will be a list containing all
of the values for the first independent variable (indicated by `0` in the independent variable node), in ascending
order, that are utilized in the two-dimensional table. Each subsequent row will have its first entry be the second
independent variable value corresponding to that row. The second entry will be the property values that map to the
independent variable combination for that row and column. The size of the lists in the second entry must be consistent
for each row. If there is a property value that does not exist for an independent variable combination, then a value of
`None` must be specified in its place. If the point including `None` is used in evaluation, an error will be raised.

Troubleshooting
---------------

> MatProps includes significant error checking to avoid invalid states or returning inaccurate data. Any error found

Reviewer: Should it be capitalized? matProps or `matProps` maybe?

Author: Fixed

raises an `Exception` that includes a helpful message. The four broad categories of matProps errors are:

* Invalid file format
* Failed node value check
* Property evaluation problem
* Problem loading or retrieving material data

The first two categories of problems are formatting errors or invalid values in the material YAML file. These problems
can only be resolved by making an edit to the material YAML file. Property evaluation issues arise from users providing

> invalid input to property functions. For instance, ask for the density outside the defined valid range of temperatures.

Reviewer: ```suggestion
invalid input to property functions. For instance, if the density is requested outside the defined valid range of temperatures.
```

Finally, problems loading or retrieving material data usually arise by trying to either load a material file (or

> directory) that does not exist or by an attempted overwrite of a material has already been saved.

Reviewer: ```suggestion
directory) that does not exist or by an attempted overwrite of a material that has already been saved.
```

    defProp("fuzz", "fuzziness", "1/m^2")
    defProp("goo", "gooiness", "m^2/s")
    defProp("squish", "squishiness", "1/Pa")

    armi.matProps.loadSafe("path/to/hilarious/materials/")

Tabulated Data
""""""""""""""
A material property can be defined not just by math, but by tabulated data. This is flexible, easy to use, and can
better match laboratory measurements. Currently, matProps supports one dimensional and two dimensional tables of data,
as shown in the example at the beginning of this section. Notice that the "type" field is used to make this distinction.

These tables of data can be large and inconvienent to duplicate. So you can tag and name the tables of data for use
elsewhere in the YAML file. (Look for "tabulated data: &tagged_data" in the example file above.)

Functions
"""""""""
The `function` field allows the developer to define a mathematical curve for a material property. For instance, in the
example file above, a function is used to represent the "Young's modulus" of the material. The independent variable "T"
is defined to represent Temperature in degrees C, and the min and max values of 0 and 700 are given. This defines the

> valid bounds upon which the "equation" can be evaluated. And the "equation" listed is a polynomial equation.

Reviewer: ```suggestion
valid bounds upon which the polynomial equation can be evaluated.
```

Perhaps I'm missing some context here?
